scripts/base/frameworks/files/main.zeek

```diff
@@ -572,5 +563,28 @@ event file_state_remove(f: fa_file) &priority=10
 
 event file_state_remove(f: fa_file) &priority=-10
 	{
-	Log::write(Files::LOG, f$info);
+	# No network connection for this file? Just write it out once without
+	# uid and c$id fields.
+	if ( ! f?$conns || |f$conns| == 0 )
+		{
+		Log::write(Files::LOG, f$info);
+		return;
+		}
+
+	# If f was seen over multiple connections, unroll them here as
+	# multiple files.log entries. In previous versions of Zeek, there
+	# would only be a single files.log entry (per worker) with multiple
+	# tx_hosts, rx_hosts and conn_uids associated. This changed with v5.1
+	# to have individual log entries that all share the same fuid value.
+	for ( [cid], c in f$conns )
+		{
+		# Make a copy of the record when there's more than one
+		# connection so that the log_files event doesn't see
+		# the same record multiple times due to it being queued
+		# by reference in Log::write() rather than by copy.
+		local info = |f$conns| > 1 ? copy(f$info) : f$info;
+		info$uid = c$uid;
+		info$id = cid;
+		Log::write(Files::LOG, info);
```

Comment:
A semantic question: by just copying the other fields, seems we can end up with information inside a single entry that comes from multiple connections. For example, for the range requests, we still see the aggregate size information. I don't think there's a way around that with a script-only implementation, it would require C++ side changes to fix that. I'm wondering if that's something to just accept for now until we make internal changes, or if it's too misleading/confusing.

Reply:
> I don't think there's a way around that with a script-only implementation, it would require C++ side changes to fix that. I'm wondering if that's something to just accept for now until we make internal changes, or if it's too misleading/confusing.

That is a good point I hadn't quite considered. The existing documentation for the `_bytes` fields are in relation to file rather than the connection, so could frame it as "redundant / replicated" information and accept it for now.

After a bit of pondering: I believe it's mostly `seen_bytes` and `missing_bytes` that should be tracked per connection. `total_bytes` should be the same anyway and `overflow_bytes` is coming from the `FileReassembler` where there's only a single one (per worker).

I can give tracking seen/missed bytes per connection a spin, I'm not sure what it would look like, but doesn't seem it should be too hard. Yes/no?

Reply:
Well, the thing is that we'd still remain inconsistent as long as *some* pieces are coming from multiple connections: at the very least, that won't yield consistent results across multiple workers; which is one thing we were hoping to fix here. To do that, I think we'd need to completely separate file state by connection. 

But not saying we should go there, nor that this is a blocker; I think we can indeed do a lot with documentation here.  

Turning just some fields into per-connection values seems confusing, too, though. And if, say, `seen_bytes` now is per-connection, we don't have the total seen value anymore either.

I think I'm in favor on going with this as is for now, just making sure the documentation is clear what's going on. That'll keep current semantics while still addressing the original problem of `files.log` being hard to parse externally.